sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -714,6 +714,17 @@ object TypeCoercion {
         // try to find the first one we can implicitly cast.
         case (_, TypeCollection(types)) => types.flatMap(implicitCast(e, _)).headOption.orNull
 
+        case (ArrayType(_, nullable), ArrayType(internalType: DataType, expectedNullable))
+          if (expectedNullable || nullable == expectedNullable) =>
```

Comment:
We should also consider the case forcing the casting result is null. Please see https://github.com/apache/spark/blob/7eb2ca8e338e04034a662920261e028f56b07395/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L94